src/cmd/display.rs

```diff
@@ -0,0 +1,108 @@
+use crate::{context::Context, path::VirtualPathBuf};
+use anyhow::Result;
+use clap::{command, Parser};
+use futures::poll;
+use futures_util::stream::StreamExt;
+use lighthouse_client::protocol::{Frame, Model, Value, Verb};
+use ratatui::{
+    backend::CrosstermBackend,
+    crossterm::{
+        event::{self, Event, KeyCode, KeyModifiers},
+        terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
+        ExecutableCommand,
+    },
+    style::Color,
+    symbols::Marker,
+    widgets::{
+        canvas::{Canvas, Rectangle},
+        Widget,
+    },
+    Terminal,
+};
+use std::{io::stdout, task::Poll, time::Duration};
+
+// Timeout for keyboard event polling.
+// This limits the execution speed of the loop
+// and also the maximum possible framerate.
+const POLL_TIMEOUT_MS: u64 = 10;
+
+#[derive(Parser)]
+#[command(bin_name = "display")]
+struct Args {
+    #[arg(
+        default_value = ".",
+        help = "The resource to display as an image stream"
+    )]
+    path: VirtualPathBuf,
+}
+
+pub async fn invoke(args: &[String], ctx: &mut Context) -> Result<String> {
+    let args = Args::try_parse_from(args)?;
+    let path = ctx.cwd.join(args.path);
+
+    let mut stream = ctx.lh.stream::<Value, Model>(&path.as_lh_vec(), Value::Nil).await?;
+
+    enable_raw_mode()?;
+    stdout().execute(EnterAlternateScreen)?;
+
+    let mut terminal = Terminal::new(CrosstermBackend::new(stdout()))?;
+    loop {
+        if let Poll::Ready(Some(Ok(msg))) = poll!(stream.next()) {
```

Comment:
This is an interesting way to go about it. Feels a bit unusual to poll a future manually in an async context (this is only very rarely done in high-level async code), but I'm not sure whether there is a better alternative.

Another option that I could think of, off top of my head, would be to spawn a separate task that runs on receive loop on the stream and forwards the messages into a channel/queue that we poll here. Something like (pseudocode, no idea whether this compiles):

```rust
let (sender, receiver) = channel();

tokio::spawn(async move {
  while let Some(msg) = stream.next().await {
    sender.send(msg);
  }
});

loop {
  if let Ok(msg) = receiver.try_recv() {
    ...
  }
}
```

Perhaps we would have to use either some form of signalling mechanism or store the handle to the spawned task to cancel it (the stream) once the user exits the command.

Reply:
The problem is that we need to wait for either a new image from the stream or a keyboard event from the ratatui library in order to exit the display. The current solution is polling both in a busy loop that is only limited by the keyboard event polling timeout.

Maybe we could spawn two tasks that poll the images and events and forwards them to a single stream that is then awaited in the main task.
Or is there something like the "select" keyword in Go, which allows you to receive from multiple channels simultaneously and process the one that has a value first?
I haven't worked with tokio and channels in Rust before.

Reply:
There's a `select!` macro: https://rust-lang.github.io/async-book/06_multiple_futures/03_select.html

I haven't used it yet, but might be worth a try.